File: web/evidence-policy.js
(function(root) {
  'use strict';
  function url(value) {
    try {
      const u = new URL(value);
      if (!['http:', 'https:'].includes(u.protocol) || u.username || u.password) return '';
      u.hash = '';
      if (/^(www\.)?(x\.com|twitter\.com)$/.test(u.hostname)) {
        const match = u.pathname.match(/^\/[^/]+\/status(?:es)?\/(\d+)\/?$/);
        if (match) return 'https://x.com/i/status/' + match[1];
      }
      return u.href.replace(/\/$/, '');
    } catch (_) { return ''; }
  }
  function status(signal) {
    if (signal.evidence_policy_version !== 1 || !Array.isArray(signal.evidence)) return 'unverified';
    const records = signal.evidence.filter(e => e && url(e.original_url) && e.confirmation_state === 'reviewed_support' && e.reviewer && e.independence_group && !e.correction_url);
    const groups = new Set(records.map(e => e.independence_group));
    return groups.size < 2 ? 'unverified' : records.some(e => e.primary === true) ? 'verified' : 'corroborated';
  }
  function saudi(s) {
    return /\b(saudi|ksa|riyadh|jeddah|neom)\b|السعودي|الرياض|جدة|نيوم/i.test(
      ['title','content','text','reason','brief_en','brief_ar','country','region','topic','market'].map(k => s[k] || '').join(' '));
  }
  function normalize(s) {
    const sourceURL = url(s.url || s.source_url || s.primary_source_url);
    let cap = 75;
    try {
      const u = new URL(s.url || s.source_url || s.primary_source_url);
      const host = u.hostname.replace(/^www\./, '');
      if (['x.com','twitter.com'].includes(host)) cap = /^\/grok\//i.test(u.pathname) || s.is_reply || s.in_reply_to_status_id ? 20 : 60;
      else if (/(\.gov|\.gov\.sa|\.edu)$/.test(host) || ['reuters.com','apnews.com','arxiv.org'].includes(host)) cap = 100;
    } catch (_) { /* Missing source stays unverified. */ }
    const base = Number(s.relevance_score ?? s.score ?? 0);
    return {...s, url: sourceURL || '#', primary_source_url: url(s.primary_source_url),
      sourcePublishedAt: s.event_time || s.sourcePublishedAt || s.source_published_at || s.published_at || s.createdAt,
      content: s.content || s.text || s.reason || '', evidence_status: status(s),
      score: Math.min(cap, Math.max(0, Number.isFinite(base) ? base : 0)), saudi_relevant: saudi(s)};
  }
  function combine(...feeds) {
    const seen = new Set(), result = [];
    for (const s of feeds.flat()) {
      if (!s || typeof s !== 'object' || s.duplicate_of) continue;
      const key = url(s.url || s.source_url || s.primary_source_url) || s.id || s.candidate_id;
      if (key && seen.has(key)) continue;
      if (key) seen.add(key);
      result.push(normalize(s));
    }
    return result;
  }
  root.AIbriefEvidence = {url, status, saudi, normalize, combine};
})(typeof window !== 'undefined' ? window : globalThis);

File: aibrief/evidence_policy.py
"""Publication policy. Collector/model fields never grant editorial verification."""
from __future__ import annotations

import hashlib
import json
import math
import re
from datetime import datetime, timezone
from pathlib import Path
from urllib.parse import urlsplit, urlunsplit

POLICY_VERSION = 1
REVIEW_PATH = Path(__file__).resolve().parents[1] / "config" / "editorial_reviews.json"


def safe_url(value):
    try:
        url = urlsplit(str(value or ""))
        if url.scheme not in {"http", "https"} or not url.hostname or url.username or url.password:
            return ""
        host = url.hostname.lower().removeprefix("www.")
        if host in {"twitter.com", "x.com"}:
            host = "x.com"
            match = re.fullmatch(r"/[^/]+/status(?:es)?/(\d+)/?", url.path)
            if match:
                return f"https://x.com/i/status/{match[1]}"
        return urlunsplit((url.scheme, url.netloc.lower(), url.path.rstrip("/"), url.query, ""))
    except ValueError:
        return ""


def claim_key(signal):
    # Bind reviews to the exact claim wording AND original URL, never entity similarity.
    title = " ".join(str(signal.get("title") or signal.get("text") or "").split())
    url = safe_url(signal.get("url") or signal.get("source_url") or signal.get("primary_source_url"))
    return hashlib.sha256((url + "\n" + title).encode()).hexdigest()


def load_reviews():
    """Only this separately reviewed local file is trusted, never source payloads."""
    try:
        data = json.loads(REVIEW_PATH.read_text(encoding="utf-8"))
        return data if isinstance(data, list) else []
    except (OSError, ValueError):
        return []


def parse_time(value):
    try:
        parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
        return parsed.astimezone(timezone.utc) if parsed.tzinfo else None
    except (TypeError, ValueError):
        return None


def freshness(event_time, observation_time, now, max_age_hours=48):
    event, observed, current = map(parse_time, (event_time, observation_time, now))
    if not event or not observed or not current:
        return "unknown"
    if (event - current).total_seconds() > 600 or (observed - current).total_seconds() > 600 or (event - observed).total_seconds() > 600:
        return "invalid"
    return "stale" if max((current - event).total_seconds(), (current - observed).total_seconds()) > max_age_hours * 3600 else "fresh"


def is_saudi(signal):
    text = " ".join(str(signal.get(k) or "") for k in (
        "title", "content", "text", "reason", "brief_en", "brief_ar", "country", "region", "topic", "market"))
    return bool(re.search(r"\b(saudi|ksa|riyadh|jeddah|neom)\b|السعودي|الرياض|جدة|نيوم", text, re.I))


def source_tier(signal):
    try:
        url = urlsplit(str(signal.get("url") or signal.get("source_url") or ""))
        host = (url.hostname or "").lower().removeprefix("www.")
        if host in {"x.com", "twitter.com"}:
            return "commentary" if url.path.lower().startswith("/grok/") or signal.get("is_reply") or signal.get("in_reply_to_status_id") else "social"
        if host.endswith((".gov", ".gov.sa", ".edu")):
            return "institution"
        return "original" if host in {"reuters.com", "apnews.com", "arxiv.org"} else "unknown"
    except ValueError:
        return "unknown"


def normalize_signal(signal, *, reviews=(), observed_at=None):
    result = dict(signal)
    key = claim_key(signal)
    original = safe_url(signal.get("url") or signal.get("source_url") or signal.get("primary_source_url"))
    urls = [original]
    for field in ("source_urls", "evidence_urls"):
        if isinstance(signal.get(field), list):
            urls.extend(safe_url(url) for url in signal[field])
    urls = list(dict.fromkeys(url for url in urls if url))
    observed = signal.get("observed_at") or signal.get("shown_at") or signal.get("seen_at") or observed_at
    event = signal.get("event_time") or signal.get("source_published_at") or signal.get("sourcePublishedAt") or signal.get("published_at") or signal.get("createdAt")
    records, groups, primary = [], set(), False
    for url in urls:
        review = next((r for r in reviews if isinstance(r, dict) and r.get("claim_key") == key and safe_url(r.get("url")) == url), {})
        group = str(review.get("independence_group") or "").strip()
        reviewed = bool(review.get("reviewer") and parse_time(review.get("reviewed_at")))
        correction = safe_url(review.get("correction_url")) or None
        supports = reviewed and review.get("supports_claim") is True and bool(group) and not correction
        if supports:
            groups.add(group)
            primary |= review.get("primary") is True
        records.append({"original_url": url, "publisher": urlsplit(url).hostname,
                        "observation_time": observed if url == original else review.get("observation_time"),
                        "event_time": event if url == original else review.get("event_time"),
                        "confirmation_state": "reviewed_support" if supports else "unverified",
                        "independence_group": group if supports else None,
                        "reviewer": review.get("reviewer") if reviewed else None,
                        "primary": bool(supports and review.get("primary") is True),
                        "correction_url": correction})
    status = ("verified" if primary else "corroborated") if len(groups) >= 2 else "unverified"
    result.update(claim_key=key, evidence=records, evidence_status=status,
                  evidence_policy_version=POLICY_VERSION, observed_at=observed,
                  event_time=event, saudi_relevant=is_saudi(signal), source_tier=source_tier(signal))
    result["freshness"] = freshness(event, observed, observed_at or observed)
    # Preserve workflow status (held/approved/etc.) separately from claim status.
    try:
        score = float(signal.get("relevance_score", signal.get("score", 0)) or 0)
    except (TypeError, ValueError):
        score = 0
    score = score if math.isfinite(score) else 0
    cap = {"commentary": 20, "social": 60, "unknown": 75}.get(result["source_tier"], 100)
    result["relevance_score"] = score
    result["score"] = min(max(score, 0), cap)
    return result
